rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+	// Logger for the subgraph request
+	Logger *zap.Logger
+}
+
+type SubgraphResponse struct {
+	// The active subgraph
+	Subgraph *core.Subgraph
+	// The original Gateway request
+	GatewayRequest *core.GatewayRequest
+	// The active Telemetry instance
+	Telemetry *core.Telemetry
+	// The final GraphQL response from the subgraph
+	Response *core.GraphQLResponse
+	// The original http response from the subgraph
+	Orignal *http.Response
+}
+
+type ModuleHooks interface {
+	// Provision is called when the module is provisioned
+	Provision(ctx *core.ModuleContext) error
+	// Shutdown is called when the module is shutdown
+	Shutdown() error
+	// Module returns the module information and factory function
+	Module() core.ModuleInfo
+}
+
+// SubgraphHooks are called when a subgraph request or response is made.
+// The order is not guaranteed, so the hooks should be idempotent and side-effect free.
+// if state needs to be shared between hooks, it should be stored in the context.
+// We will provide an easy way to share state between hooks.
+type SubgraphHooks interface {
+	// OnRequest is called when a subgraph request is made.
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphRequest(req *core.SubgraphRequest, err error) error
+	// OnResponse is called when a subgraph response is received
+	// Returning an error will result in a GraphQL error as a response from the subgraph.
+	OnSubgraphResponse(res *core.SubgraphResponse, err error) error
+	// OnError is called when an error occurs during the subgraph lifecycle request/response or any other error
+	OnError(err error)
+}
+
+type ApplicationHooks interface {
+	// OnStart is called when the application starts
+	OnAppStart() error
+	// OnStop is called when the application stops
+	OnAppStop() error
+	// OnError is called when an error occurs during the application lifecycle start/stop or any other error
+	OnAppError(err error)
+}
+
+type TelemetryHooks interface {
+	// OnSpanStart is called when a span is created
+	// Returning a function to be called when the span ends.
+	// This can be used to add custom attributes or events to the span.
+	OnSpan(span *trace.Span) func() // Return a function to be called when the span ends
+}
+
+type AuthenticationHooks interface {
+	// OnAuthenticate is called when a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnAuthenticate(req *core.GatewayRequest, err error) error
+	// OnPostAuthenticate is called after a gateway request is authenticated
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnPostAuthenticate(req *core.GatewayRequest, err error) error
+}
+
+type OperationHooks interface {
+	// OnNormalize is called when an operation is normalized
+	// Returning an error will result in a GraphQL error as a response from the gateway.
+	OnOperationNormalize(op *core.Operation, err error) error
```

Comment:
Should we follow a pattern of "Pre" + "Post", just to be clear when it's being called?
This question applies to all occurrences of On... and OnPost...

Reply:
I added a note to the RFC that we will evaluate Pre/Post hooks for certain methods. I don't think it makes sense for all. A strong point for Pre/Post hooks is tracing, where users can precisely measure the time of a hook themselves, but this would require that they can disable built-in tracing otherwise we would generate duplicates.

Reply:
+1 to "Pre" + "Post". It's clearer what time point they're called. 

IMO, tracing should be just another plugin, a default built-in plugin that's built on top of this RFC. They should by default off. User should either implement their own tracing plugin or enable the default.